ITS: Fix GPU deterministic mode + refactoring

f3sch opened the pull request:
Fixes two things for ITS GPU to be deterministic again.
https://github.com/AliceO2Group/AliceO2/blob/8c94458e83debdbc773d4e1a57923a6e63733c2d/Detectors/ITSMFT/ITS/tracking/GPU/cuda/TrackingKernels.cu#L87
Takes ZBin which is constexpr 256 but should be taken from params which set it to 64.
https://github.com/AliceO2Group/AliceO2/blob/8c94458e83debdbc773d4e1a57923a6e63733c2d/Detectors/ITSMFT/ITS/tracking/GPU/cuda/TrackingKernels.cu#L1285
Should be `foundSeedsTable.size()`, since we can find more seeds than `nCells`.
with this I get (although there still are some bins with ~0.01% difference in 10 pp TFs. Any idea why? 
![ratio](https://github.com/user-attachments/assets/af46fe6b-7b3c-45f4-a02f-766fa3e0c92a)

The rest is mostly a refactoring which mostly happened on-the-fly especially the `clusterToTracksImpl` since I wanted to be sure that both algorithms do not diverge at any point.

alibuild commented:
Error while checking build/O2/fullCI_slc9 for 8be3ee504d0c6f768346aede0f8a6ea505649b18 at 2025-04-17 22:43:
```
## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
/sw/BUILD/11921aa24f97e95c080dfce14a35f98fd40639b9/O2/Detectors/ITSMFT/ITS/tracking/GPU/hip/TrackingKernels.hip:802:85: error: use nullptr [modernize-use-nullptr]
/sw/BUILD/11921aa24f97e95c080dfce14a35f98fd40639b9/O2/Detectors/ITSMFT/ITS/tracking/GPU/hip/TrackingKernels.hip:815:90: error: use nullptr [modernize-use-nullptr]
/sw/BUILD/11921aa24f97e95c080dfce14a35f98fd40639b9/O2/Detectors/ITSMFT/ITS/tracking/GPU/hip/TrackingKernels.hip:821:85: error: use nullptr [modernize-use-nullptr]
/sw/BUILD/11921aa24f97e95c080dfce14a35f98fd40639b9/O2/Detectors/ITSMFT/ITS/tracking/GPU/hip/TrackingKernels.hip:834:90: error: use nullptr [modernize-use-nullptr]
/sw/BUILD/11921aa24f97e95c080dfce14a35f98fd40639b9/O2/Detectors/ITSMFT/ITS/tracking/GPU/hip/TrackingKernels.hip:898:84: error: use nullptr [modernize-use-nullptr]
++ [[ 0 == 0 ]]
++ exit 1
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14200/8be3ee504d0c6f768346aede0f8a6ea505649b18/build_O2_fullCI_slc9/pretty.html).

alibuild commented:
Error while checking build/O2/fullCI_slc9 for 55f819d49652144c7eafe3390f248bc1d04a2689 at 2025-04-18 12:04:
```
## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
/sw/BUILD/11921aa24f97e95c080dfce14a35f98fd40639b9/O2/Detectors/ITSMFT/ITS/tracking/GPU/hip/TrackingKernels.hip:802:85: error: use nullptr [modernize-use-nullptr]
/sw/BUILD/11921aa24f97e95c080dfce14a35f98fd40639b9/O2/Detectors/ITSMFT/ITS/tracking/GPU/hip/TrackingKernels.hip:815:90: error: use nullptr [modernize-use-nullptr]
/sw/BUILD/11921aa24f97e95c080dfce14a35f98fd40639b9/O2/Detectors/ITSMFT/ITS/tracking/GPU/hip/TrackingKernels.hip:821:85: error: use nullptr [modernize-use-nullptr]
/sw/BUILD/11921aa24f97e95c080dfce14a35f98fd40639b9/O2/Detectors/ITSMFT/ITS/tracking/GPU/hip/TrackingKernels.hip:834:90: error: use nullptr [modernize-use-nullptr]
/sw/BUILD/11921aa24f97e95c080dfce14a35f98fd40639b9/O2/Detectors/ITSMFT/ITS/tracking/GPU/hip/TrackingKernels.hip:898:84: error: use nullptr [modernize-use-nullptr]
++ [[ 0 == 0 ]]
++ exit 1
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14200/55f819d49652144c7eafe3390f248bc1d04a2689/build_O2_fullCI_slc9/pretty.html).

mconcas commented:
Hi, thanks for looking into this. As per usual it would be better to separate code refactoring from small fixes. Can you split it in two?

f3sch committed:
ITS: fix GPU tracking

- compute-sanitizer reveal malicious write past allocated table
- ZBins for map lookup was not taken from params

f3sch committed:
ITS: GPU refactoring

- Removes unused headers
- adds two new functions for ex/in-clusive scan via cub
- inlines square
- applies deterministc mode blocks=1,threads=1 to all kernels

f3sch committed:
ITS: unify hybrid/cpu clusterToTracks interface

f3sch commented:
Sure, now first commit is the actual fix, second the refactoring work done in GPU kernel code, the third is the refactoring of the hybrid/cpu call chain.

f3sch committed:
ITS: drop hybrid functions use dynamic dispatch

mconcas commented on Detectors/ITSMFT/ITS/tracking/src/Tracker.cxx:
```diff
@@ -105,131 +100,52 @@ void Tracker::clustersToTracks(std::function<void(std::string s)> logger, std::f
-    logger(fmt::format(" - Track finding: {} tracks found in {:.2f} ms", nTracks + mTimeFrame->getNumberOfTracks(), timeRoads));
+    logger(std::format(" - Tracklet finding: {} tracklets found in {:.2f} ms", nTracklets, timeTracklets));
+    logger(std::format(" - Cell finding: {} cells found in {:.2f} ms", nCells, timeCells));
+    logger(std::format(" - Meighbours finding: {} neighbours found in {:.2f} ms", nNeighbours, timeNeighbours));
```
```suggestion
    logger(std::format(" - Neighbours finding: {} neighbours found in {:.2f} ms", nNeighbours, timeNeighbours));
```

f3sch commented on dependencies/FindO2GPU.cmake:
```diff
@@ -68,7 +68,7 @@ endfunction()
 
 # ---------------------------------- Fast Math / Deterministic Mode ----------------------------------
-# set(GPUCA_DETERMINISTIC_MODE WHOLEO2)          # Override
+set(GPUCA_DETERMINISTIC_MODE WHOLEO2)          # Override
```
N.B. this will of course not go into the PR :)

f3sch committed:
ITS: add total/avg summary + remove some headers

mconcas committed:
Make ITS GPU tracking deterministic again

f3sch commented:
Now it is up to the precision of TH1F identical :)
20 TFs in both of high IR pp.
<img width="1191" alt="image" src="https://github.com/user-attachments/assets/d970d06a-2eec-459c-97d0-1e763aeb1975" />
